[Security] Deprecated supportsAttribute and supportsClass methods

wouterj opened the pull request:
These methods aren't used at all in a Symfony application and don't make sense to use in the application. They are only used internally in the voters. This means the voter interface can be made much easier.

I'm not sure how we do these deprecations, should we remove the methods from the interface now already? Also, I don't think it's possible to trigger deprecation notices for the voter methods?

| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | one of #11742 |
| License | MIT |
| Doc PR | - |
## Abstract Voter

There is one remaining question about the abstract voter. This currently has abstract `getSupportedAttributes()` and `getSupportedClass()` methods. One of the reasons to remove the methods for the interface was that these methods are not flexible. Does it make sense to deprecate these methods as well and replace them by an abstract `protected vote(array $attributes, $class)` method in the `AbstractVoter` (which is called from `AbstractVoter#vote()`) ?

wouterj committed:
Deprecated supportsAttribute and supportsClass methods

dosten commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php:
```diff
@@ -77,6 +77,8 @@ public function decide(TokenInterface $token, array $attributes, $object = null)
      */
     public function supportsAttribute($attribute)
     {
+        trigger_error('The '.__METHOD__.' is deprecated since version 2.8 and will be removed in version 3.0.');
```
You should append the `trigger_error` call with `@`

dosten commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php:
```diff
@@ -91,6 +93,8 @@ public function supportsAttribute($attribute)
      */
     public function supportsClass($class)
     {
+        trigger_error('The '.__METHOD__.' is deprecated since version 2.8 and will be removed in version 3.0.');
```
same here

dosten commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManagerInterface.php:
```diff
@@ -37,6 +37,8 @@ public function decide(TokenInterface $token, array $attributes, $object = null)
      *
      * @return bool true if this decision manager supports the attribute, false otherwise
+     *
+     * @deprecated Since version 2.8, to be removed in 3.0.
```
AFAIK we do not uppercase the first letter here, should be `@deprecated since version 2.8, to be removed in 3.0.`

dosten commented on src/Symfony/Component/Security/Core/Authorization/AccessDecisionManagerInterface.php:
```diff
@@ -46,6 +48,8 @@ public function supportsAttribute($attribute);
      *
      * @return true if this decision manager can process the class
+     *
+     * @deprecated Since version 2.8, to be removed in 3.0.
```
same here

dosten commented on src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php:
```diff
@@ -30,6 +30,8 @@
      *
      * @return bool true if this Voter supports the attribute, false otherwise
+     *
+     * @deprecated Since version 2.8, to be removed in 3.0.
```
same here

dosten commented on src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php:
```diff
@@ -39,6 +41,8 @@ public function supportsAttribute($attribute);
      *
      * @return bool true if this Voter can process the class
+     *
+     * @deprecated Since version 2.8, to be removed in 3.0.
```
same here

WouterJ commented:
Thanks @dosten. Fixed everything you mentioned :)

stof commented:
@WouterJ I suggest you to mark methods as deprecated in the VoterInterface too, and to trigger the deprecation in the AbstractVoter when they are called.
We should not trigger deprecations for people implementing the method: it is required, and it would not hurt to have them on 3.0 anyway.

Soullivaneuh commented:
> These methods aren't used at all in a Symfony application and don't make sense to use in the application. 

What do you mean? I use those methods on my Voters class, working with it.

Sample: 

```php
class DomainVoter extends AbstractVoter
{
    const VIEW      = 'view';
    const EDIT      = 'edit';
    const DELETE    = 'delete';

    /**
     * {@inheritdoc}
     */
    protected function getSupportedClasses()
    {
        return [
            'AppBundle\Entity\Domain',
        ];
    }

    /**
     * {@inheritdoc}
     */
    protected function getSupportedAttributes()
    {
        return [
            self::VIEW,
            self::EDIT,
            self::DELETE,
        ];
    }

    /**
     * @param string               $attribute
     * @param Domain               $object
     * @param UserInterface|string $user
     *
     * @return bool
     */
    protected function isGranted($attribute, $object, $user = null)
    {
        return true; // Stuff
    }
}
```

What should be done instead of using those deprecated methods?

stof commented:
@Soullivaneuh if you look at the AbstractVoter, it uses these methods internally. But the methods of the interface are never used.
The AbstractVoter could keep using these methods (to hook into the main logic of the ``vote`` method), but they should not be part of the VoterInterface

stof commented:
And btw, the methods you are showing there are **not** the methods which are meant to be deprecated

fabpot commented:
@WouterJ Can you finish this one?

Soullivaneuh commented:
@stof Indeed, my bad. Wrong reading.

wouterj committed:
Rewrote logic of AbstractVoter

WouterJ commented:
Updated this PR:

* I've changed the API of `AbstractVoter` to have a `supports($attribute, $class)` method. The reason behind this is to allow some attributes to be accepted for only some classes.
* In order to remove too much duplication in `AbstractVoter#supports()`, I've added an `AbstractVoter#isClassInstanceOf()` helper method.
* I've added some tests for `AbstractVoter`, in order to make sure the behaviour is kept equal.
* I tried to trigger deprecation in as much places as possible.

Fabbot's error is a bug in fabbot.

fabpot commented:
:+1: (should be rebased)

fabpot commented:
ping @symfony/deciders

Taluu commented on src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php:
```diff
@@ -83,18 +88,76 @@ public function vote(TokenInterface $token, $object, array $attributes)
+     */
+    protected function supports($attribute, $class)
+    {
+        @trigger_error('The getSupportedClasses and getSupportedAttributes methods are deprecated since version 2.8 and will be removed in version 3.0. Overwrite supports instead.');
```
why this trigger here ?

WouterJ replied:
People have to override the supports method to see if the attribute and class are supported. If they didn't, this code will be executed and they will recieve a deprecation notice. If they did, this code is overriden and the deprecation notice isn't triggered.

The triggers in the `getSupported*` classes are there in case someone is executing this methods directly, to indicate that they have to use `supports`.

Taluu replied:
Well, if this method is executed, then wouldn't the other deprecations also be triggered ? So basically here we will have at least 2 (and at most 3) deprecations warnings (one for the call on `supports`, which triggers this as BC, and one or two for the usage of `getSupported*` methods)

WouterJ replied:
Yes, but if we remove one of them, we'll loose information or we didn't cover all BC breaks. There are more cases in which you get 2 or more deprecation messages by using 1 deprecated feature (every deprecated setting configurable in yaml/xml/php for example)

Taluu replied:
Getting the 2 deprecation messages for the deprecated calls to `getSupported*` are fine IMO, but I think the deprecation warning for this method is not necessary, that was my point.

WouterJ commented:
Rebased, so it can be merged directly if accepted by core

Taluu commented on src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php:
```diff
@@ -96,18 +101,76 @@ public function vote(TokenInterface $token, $object, array $attributes)
+     *
+     * @return bool True if the attribute and class is supported, false otherwise
+     */
+    protected function supports($attribute, $class)
```
I guess this method will turn abstract on 3.0 ?

stof replied:
yes it will. And this is why there is a deprecation warning in the provided implementation.

@WouterJ please add a comment in the phpdoc saying it will become abstract in 3.0

WouterJ replied:
added

xabbuh commented:
Seems like some tests still test the deprecated features and should be added to the legacy group (:+1: when this is fixed).

wouterj committed:
Cleanup AbstractVoter tests

wouterj committed:
Indicate method will become abstract in 3.0

WouterJ commented:
So it seems like we now had 2 tests for AbstractVoter (one in `Security\Tests\Core` and one in `Security\Core\Tests`). I've now merged both test classes, everything should work now

xabbuh commented:
:+1: (the failures don't seem to be related)

Status: Reviewed

fabpot commented:
Thank you @WouterJ.

weaverryan commented:
@WouterJ All the `trigger_error()` calls in this PR are missing the `E_USER_DEPRECATED` flag at the end (so they're actually notices right now).

WouterJ commented:
Whoa, thanks for catching this @weaverryan! Fixed in https://github.com/symfony/symfony/pull/15922
